staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject.go

```diff
@@ -118,13 +119,14 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 					return nil, errors.New(err)
 				}
 				containerName = t.Spec.Containers[0].Name
-			} else {
-				containerName = opts.Container
+			}
+			if opts != nil {
+				opts.Container = containerName
```

Comment:
🤔 why do we only need to set the container name into options when we had options passed in? 

There are three possible sources of a container name, and which of them set `containerName` and `opts.Container` is now changing in this PR:
- `opts.Container`: both are set before and after
- annotations: previously both were set, now `opts.Container` is not set if `opts` as a whole is nil.
- sole container: only `containerName` was set before, and both will be set after if `opts` is non-nil

Since `opts` (and not `containerName`) is what is given to the client, I would expect `opts.ContainerName` to always need to be set when we're selecting a specific container. It's a bit complicated, and I don't know if `opts` is actually ever `nil` in production, but I can produce a test failure as follows:
1. Change `opts: &corev1.PodLogOptions{}` to `opts: nil` in your test case
2. Add a second container to pod "foo" in the test case (so that the "sole container" case won't apply)
3. Fix a segfault at L83 in this file that (2) exposes
4. Run the test: `pods list logs: different containers: unexpected error: a container name must be specified for pod foo, choose one of: [c1 c2]`

My reaction is that it is odd that the sole container case did not set `opts.Container` in the first place--it's doing all the work to get and verify the name, but not using it in the request. I'd be inclined to simplify the code by creating a context-specific deep copy and using it instead of containerName as well (see diff below the fold at the end of this comment). That makes some tests fail that are expecting nil options, but they pass if I make them instead expect to receive options that include the correct container name. @eddiezane @soltysh WDYT?

<details>

<summary>Suggested diff</summary>

```diff
diff --git a/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject.go b/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject.go
index ef05e2f575e..7f5a275b01a 100644
--- a/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject.go
+++ b/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject.go
@@ -58,7 +58,6 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 	if !ok {
 		return nil, errors.New("provided options object is not a PodLogOptions")
 	}
-	opts = opts.DeepCopy()
 
 	switch t := object.(type) {
 	case *corev1.PodList:
@@ -77,31 +76,34 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 	case *corev1.Pod:
 		// if allContainers is true, then we're going to locate all containers and then iterate through them. At that point, "allContainers" is false
 		if !allContainers {
-			var containerName string
+			currOpts := new(corev1.PodLogOptions)
+			if opts != nil {
+				opts.DeepCopyInto(currOpts)
+			}
 			// in case the "kubectl.kubernetes.io/default-container" annotation is present, we preset the opts.Containers to default to selected
 			// container. This gives users ability to preselect the most interesting container in pod.
-			if annotations := t.GetAnnotations(); annotations != nil && len(opts.Container) == 0 {
+			if annotations := t.GetAnnotations(); annotations != nil && currOpts.Container == "" {
+				var defaultContainer string
 				if len(annotations[podcmd.DefaultContainerAnnotationName]) > 0 {
-					containerName = annotations[podcmd.DefaultContainerAnnotationName]
+					defaultContainer = annotations[podcmd.DefaultContainerAnnotationName]
 				} else if len(annotations[defaultLogsContainerAnnotationName]) > 0 {
 					// Only log deprecation if we have only the old annotation. This allows users to
 					// set both to support multiple versions of kubectl; if they are setting both
 					// they must already know it is deprecated, so we don't need to add noisy
 					// warnings.
-					containerName = annotations[defaultLogsContainerAnnotationName]
+					defaultContainer = annotations[defaultLogsContainerAnnotationName]
 					fmt.Fprintf(os.Stderr, "Using deprecated annotation `kubectl.kubernetes.io/default-logs-container` in pod/%v. Please use `kubectl.kubernetes.io/default-container` instead\n", t.Name)
 				}
-				if len(containerName) > 0 {
-					if exists, _ := podcmd.FindContainerByName(t, containerName); exists == nil {
-						fmt.Fprintf(os.Stderr, "Default container name %q not found in a pod\n", containerName)
-						containerName = ""
+				if len(defaultContainer) > 0 {
+					if exists, _ := podcmd.FindContainerByName(t, defaultContainer); exists == nil {
+						fmt.Fprintf(os.Stderr, "Default container name %q not found in pod %s\n", defaultContainer, t.Name)
+					} else {
+						currOpts.Container = defaultContainer
 					}
 				}
-			} else if opts != nil {
-				containerName = opts.Container
 			}
 
-			if opts == nil || len(containerName) == 0 {
+			if currOpts.Container == "" {
 				// We don't know container name. In this case we expect only one container to be present in the pod (ignoring InitContainers).
 				// If there is more than one container, we should return an error showing all container names.
 				if len(t.Spec.Containers) != 1 {
@@ -118,15 +120,12 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 
 					return nil, errors.New(err)
 				}
-				containerName = t.Spec.Containers[0].Name
-			}
-			if opts != nil {
-				opts.Container = containerName
+				currOpts.Container = t.Spec.Containers[0].Name
 			}
 
-			container, fieldPath := podcmd.FindContainerByName(t, containerName)
+			container, fieldPath := podcmd.FindContainerByName(t, currOpts.Container)
 			if container == nil {
-				return nil, fmt.Errorf("container %s is not valid for pod %s", containerName, t.Name)
+				return nil, fmt.Errorf("container %s is not valid for pod %s", currOpts.Container, t.Name)
 			}
 			ref, err := reference.GetPartialReference(scheme.Scheme, t, fieldPath)
 			if err != nil {
@@ -134,7 +133,7 @@ func logsForObjectWithClient(clientset corev1client.CoreV1Interface, object, opt
 			}
 
 			ret := make(map[corev1.ObjectReference]rest.ResponseWrapper, 1)
-			ret[*ref] = clientset.Pods(t.Namespace).GetLogs(t.Name, opts)
+			ret[*ref] = clientset.Pods(t.Namespace).GetLogs(t.Name, currOpts)
 			return ret, nil
 		}
 
```

</details>


<details>

<summary>Corresponding test diff</summary>

```diff
diff --git a/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject_test.go b/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject_test.go
index c9bd31316b8..65a0c75d402 100644
--- a/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject_test.go
+++ b/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/logsforobject_test.go
@@ -56,7 +56,7 @@ func TestLogsForObject(t *testing.T) {
 			name: "pod logs",
 			obj:  testPodWithOneContainers(),
 			actions: []testclient.Action{
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{
 				{
@@ -129,7 +129,7 @@ func TestLogsForObject(t *testing.T) {
 				Items: []corev1.Pod{*testPodWithOneContainers()},
 			},
 			actions: []testclient.Action{
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,
@@ -140,8 +140,7 @@ func TestLogsForObject(t *testing.T) {
 			}},
 		},
 		{
-			name: "pods list logs: different containers",
-			opts: &corev1.PodLogOptions{},
+			name: "pods list logs: default container should not leak across pods",
 			obj: &corev1.PodList{
 				Items: []corev1.Pod{
 					{
@@ -162,6 +161,7 @@ func TestLogsForObject(t *testing.T) {
 							DNSPolicy:     corev1.DNSClusterFirst,
 							Containers: []corev1.Container{
 								{Name: "c1"},
+								{Name: "c2"},
 							},
 						},
 					},
@@ -265,7 +265,7 @@ func TestLogsForObject(t *testing.T) {
 			clientsetPods: []runtime.Object{testPodWithOneContainers()},
 			actions: []testclient.Action{
 				testclient.NewListAction(podsResource, podsKind, "test", metav1.ListOptions{LabelSelector: "foo=bar"}),
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,
@@ -286,7 +286,7 @@ func TestLogsForObject(t *testing.T) {
 			clientsetPods: []runtime.Object{testPodWithOneContainers()},
 			actions: []testclient.Action{
 				testclient.NewListAction(podsResource, podsKind, "test", metav1.ListOptions{LabelSelector: "foo=bar"}),
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,
@@ -307,7 +307,7 @@ func TestLogsForObject(t *testing.T) {
 			clientsetPods: []runtime.Object{testPodWithOneContainers()},
 			actions: []testclient.Action{
 				testclient.NewListAction(podsResource, podsKind, "test", metav1.ListOptions{LabelSelector: "foo=bar"}),
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,
@@ -328,7 +328,7 @@ func TestLogsForObject(t *testing.T) {
 			clientsetPods: []runtime.Object{testPodWithOneContainers()},
 			actions: []testclient.Action{
 				testclient.NewListAction(podsResource, podsKind, "test", metav1.ListOptions{LabelSelector: "foo=bar"}),
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,
@@ -349,7 +349,7 @@ func TestLogsForObject(t *testing.T) {
 			clientsetPods: []runtime.Object{testPodWithOneContainers()},
 			actions: []testclient.Action{
 				testclient.NewListAction(podsResource, podsKind, "test", metav1.ListOptions{LabelSelector: "foo=bar"}),
-				getLogsAction("test", nil),
+				getLogsAction("test", &corev1.PodLogOptions{Container: "c1"}),
 			},
 			expectedSources: []corev1.ObjectReference{{
 				Kind:       testPodWithOneContainers().Kind,

```

</details>

Reply:
Awesome suggestion! Thank you very much!